server/server.go

```diff
@@ -59,28 +64,70 @@ func (s *Server) Run(svc transport.Service) error {
 }
 
 func (s *Server) run() error {
-	errChan := make(chan error, 5)
-
+	var g errgroup.Group
 	if s.listeners.mainListener != nil {
-		go func() {
+		g.Go(func() error {
 			err := s.listeners.mainListener.Serve()
-			errChan <- err
-		}()
+			return err
+		})
+	}
+	sigint := make(chan os.Signal, 1)
+	signal.Notify(sigint, os.Interrupt)
+	if s.opts.GracefullFunc != nil {
+		fn := s.opts.GracefullFunc(sigint)
```

Comment:
What can we use `GracefulFunc` for?

Reply:
For graceful shutdown server with custom func? No? Example will be later today

Reply:
https://github.com/karantin2020/clay/blob/beef98437a18f5604947ae5695e5acb16c03ea13/server/server.go#L80

Is an example of GracefulFunc and goes as default value. But may be it's better to call server.StopWithTimeout(value)

Reply:
Yeah, I understand how the sigs work, but why is it in clay's code? This handler should be in your own code since it's specific for you and it doesn't do anything with the server's internals. Just call GStop when you want.

Reply:
1. server.Stop() doesn't implement Gracefull shutdown at this moment (https://github.com/utrack/clay/blob/647205ee9f3163a4e52043fa44736b264ae465ef/server/server.go#L84)
2. There's no http.Server in clay's server.Server, nor in server.Server.srv  
```
type Server struct {
	opts      *serverOpts
	listeners *listenerSet
	srv       *serverSet
}
```  
That's why we cann't call http.Server.Shutdown from anywhere  

Another way is to add importable variable `http.Server HTTPServer` to caly's server.Server or anywhere alse and call it to gracefully shutdown from upper code.  

Add importable `http.Server` variable. Let users to configure it and call all it's methods.

What is your opinion?

Reply:
1) yep, and now it should stay that way. Your PR changes `Stop` behaviour to `GracefulStop(context.Background()`, that potentially hangs the server indefinitely instead of stopping right away - hence the new `GracefulStop()` func is needed.
2) `clay/server.Server` has `Stop()` and (new) `GracefulStop()`, that should be called by the user (and they, in turn, stop http and gRPC servers). There's no need to expose `http.Server.Shutdown`.

Reply:
GracefulStop() with no arguments? What it should do?

Reply:
`GracefulStop(ctx)` - context may have timeout in it.

Reply:
`Stop()` should continue calling non-graceful `Stop()` to preserve old (and anticipated) behaviour.

Reply:
What to do is user wants to implement it's own logic in GracefulStop()?

Reply:
Which logic? `GracefulStop` stops the servers w/ timeout, nothing more. What do you want to do?

Reply:
May be like `func (srv *Server) RegisterOnShutdown(f func())` was made for some logic?  
I think that it's more correct to use stdlib funcs explicitly.  
Do you think that it's possible to expose *http.Server and *grpc.Server variables?

Reply:
```
// your code
go clearWsConnections() // or whatever cleanup you want to do over hijacked conns
err := server.GracefulStop(ctx)
```
should be enough since `GracefulStop` is called by your code. 
`RegisterOnShutdown()` was originally made for obscure scenario when HTTP2 server is initiated via `http2.ConfigureServer` and http2 does not have any control over http1 server or user code. 

Even more so, this kind of hook is user-specific - server can be stopped by other means other than signal and most users would need more info about the shutdown, so your function signature would be insufficient; only you would use this func, but it would pollute the package for everyone. You should handle the cleanup in your own code since it's really easy to do so there.

Exposing http and gRPC servers would break feature parity for both servers and increase the cognitive load. 
It's a good idea to introduce some unified options that would work for both HTTP and gRPC clients like TLSConfig, ReadTimeout and *Timeout though.